cpp/memilio/utils/multi_index_range.h

```diff
@@ -0,0 +1,162 @@
+/* 
+* Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
+*
+* Authors: René Schmieding
+*
+* Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#ifndef MIO_UTILS_MULTI_INDEX_RANGE_H_
+#define MIO_UTILS_MULTI_INDEX_RANGE_H_
+
+#include "memilio/utils/index.h"
+
+namespace mio
+{
+
+/**
+ * @brief A Range that can be used to iterate over a MultiIndex.
+ * @tparam MultiIndex A type mio::Index<Categories...> for some set of categories.
+ */
+template <class MultiIndex>
+class MultiIndexRange
+{
+public:
+    /**
+     * @brief Construct a Range that can be used to iterate over all MultiIndices in the given dimensions.
+     * The range for each Index i in the MultiIndex is determined by [0, d_i), where d_i is the dimension of i.
+     * @param[in] dimensions A MultIndex that contains the dimension for each Category.
+     */
+    MultiIndexRange(const MultiIndex& dimensions)
+        : m_dimensions(dimensions)
+        , m_end([](const MultiIndex& dims) {
+            // set end to the first invalid index that is reached by increments of 1,
+            // i.e. 0 everywhere except for the most significant index (position 0),
+            // which is set to its dimension
+            auto end         = MultiIndex::Zero();
+            mio::get<0>(end) = mio::get<0>(dims);
+            return end;
+        }(dimensions))
+    {
+    }
+
+    /**
+     * @brief Iterator for MultiIndices.
+     */
+    class MultiIndexIterator
+    {
+    public:
+        using value_type = MultiIndex;
+        using reference  = const MultiIndex&;
```

Comment:
`reference` does not need to be an actual reference type. I think this should return copies since someone might keep a reference to the index and keep iterating, getting weird results.

```suggestion
        using reference  = value_type;
```